gcp_variant_transforms/beam_io/vcfio.py

```diff
@@ -207,6 +209,145 @@ def __repr__(self):
+               coder=coders.ToStringCoder(),
+               compression_type=CompressionTypes.UNCOMPRESSED,
+               headers=None):
+    """Initialize a _TextSink.
+
+    Args:
+      file_path: The file path to write to. The files written will begin
+        with this prefix, followed by a shard identifier (see num_shards), and
+        end in a common extension, if given by file_name_suffix. In most cases,
+        only this argument is specified and num_shards, shard_name_template, and
+        file_name_suffix use default values.
+      num_shards: The number of files (shards) used for output. If not set, the
+        service will decide on the optimal number of shards.
+        Constraining the number of shards is likely to reduce
+        the performance of a pipeline.  Setting this value is not recommended
+        unless you require a specific number of output files.
+      coder: Coder used to encode each line.
+      compression_type: Used to handle compressed output files. Typical value
+        for VCF files is CompressionTypes.UNCOMPRESSED. If set to
+        CompressionTypes.AUTO, file_path's extension will be used to detect
+        compression.
+      headers: A list of VCF meta-information lines describing the INFO,
+        FILTER, and FORMAT entries in each record and a header line describing
+        the column names. These lines will be written at the beginning of the
+        file.
+    """
+    super(_VcfSink, self).__init__(
+        file_path,
+        coder,
+        num_shards=num_shards,
+        mime_type='text/plain',
+        compression_type=compression_type)
+    self._headers = headers
+
+  def open(self, temp_path):
+    file_handle = super(_VcfSink, self).open(temp_path)
+    if self._headers is not None:
+      for header in self._headers:
+        file_handle.write(header)
+    return file_handle
+
+  def _get_variant_info(self, variant):
+    info_list = []
+    # Set end in info if this is a non-variant.
+    if (variant.start and variant.reference_bases and
+        variant.start + len(variant.reference_bases) < variant.end):
+      info_list.append('END=%d' % variant.end)
+    # Set all other fields of info.
+    for k, v in variant.info.iteritems():
+      if v.field_count == '1':
+        info_list.append('%s=%s' % (k, str(v.data)))
+      else:
+        info_list.append('%s=%s' % (k, ','.join([str(x) for x in v.data])))
+    return ';'.join(info_list)
+
+  def _get_variant_format(self, variant):
+    format_keys = ['GT']
+    if variant.calls:
+      # If any calls have a set phaseset that is not `DEFAULT_PHASESET_VALUE`,
+      # the key will be added to the format field.
+      for call in variant.calls:
+        if call.phaseset and call.phaseset != DEFAULT_PHASESET_VALUE:
+          format_keys.append('PS')
+          break
+      format_keys.extend([k for k in sorted(variant.calls[0].info)])
+    return ':'.join(format_keys)
+
+  def _get_variant_calls(self, variant):
```

Comment:
I think it's more robust for this method to get variant_format_keys as an argument. (you can return a list of keys instead of a string in _get_variant_format_keys). That way, you can loop through the keys for each call and fill the corresponding value. You also won't rely on the sorted order of keys.

Reply:
Done.